cartography/graph/loader.py

```diff
@@ -0,0 +1,144 @@
+import logging
+from typing import Dict
+from typing import List
+
+import neo4j
+
+from cartography.graph.querybuilder import build_node_ingestion_query
+from cartography.graph.querybuilder import build_relationship_ingestion_query
+
+logger = logging.getLogger(__name__)
+
+
+def merge_nodes(
+    neo4j_session: neo4j.Session, node_label: str, node_property_map: Dict[str, str],
+    node_data_list: List[Dict], update_tag: int, batch_size: int = 10000,
+) -> None:
+    """
+    Writes the `node_data_list` to Neo4j using the UNWIND pattern in batches with
+    just one MERGE per statement for performance and reliability. Follows guidelines
+    described in
+    - https://neo4j.com/blog/bulk-data-import-neo4j-3-0/
+    - https://dzone.com/articles/tips-for-fast-batch-updates-of-graph-structures-wi
+
+    The query run looks like
+
+    UNWIND {DictList} AS item
+        MERGE (i:`node_label`{id:item.`node_property_map['id']`})
+        ON CREATE SET i.firstseen = timestamp()
+        SET i.lastupdated = {UpdateTag},
+        ... <expand the given node property map to set the other node fields> ...
+
+    :param neo4j_session: The Neo4j session
+    :param node_label: The label of the node to MERGE to Neo4j
+    :param node_property_map: Mapping from node field names to dict keys in `node_data_list`.
+    :param node_data_list: A list of dicts to be written as nodes to the graph.
+    :param update_tag: The cartography update tag.
+    :param batch_size: Optional: the number of nodes to write in a single batch operation.
+    Default = 10000.
+    :return: None.
+    """
+    if batch_size < 1:
+        raise ValueError("batch_size cannot be less than 1.")
+
+    ingest_query = build_node_ingestion_query(node_label, node_property_map)
+
+    chunk = node_data_list[:batch_size]
+    cursor = 0
+    while len(chunk) > 0:
+        query_args = {
+            'UpdateTag': update_tag,
+            'DictList': chunk,
+            'NodeLabel': node_label,  # used for debug logs
+        }
+        neo4j_session.write_transaction(_merge_chunk, ingest_query, query_args)
+        cursor += batch_size
+        chunk = node_data_list[cursor:cursor + batch_size]
+
+
+def _merge_chunk(tx: neo4j.Transaction, ingest_query: str, query_args: Dict) -> None:
```

Comment:
@pchheda-lyft @gkleiman This might be a good place to emit stats since this is eventually where all writes will be performed. Can you have a skim and let me know your thoughts?